src/utils.h

```diff
@@ -0,0 +1,148 @@
+double ColNormsMax(const T& x) {
+  return arma::sum(arma::square(x), 0).max();
+}
+
+//' Retrieve step size for the chosen family and feature matrix X
+//'
+//' @param max_squared_sum the maximum squared rowise norm from X
+//' @param alpha_scaled alpha/n_samples
+//' @param family the model family
+//' @param fit_intercept should an intercept be added to the model?
+//' @param n_samples the number of samples in X
+//'
+//' @return Step size.
+//'
+//' @keywords internal
+double GetStepSize(double         max_squared_sum,
+                   double         alpha_scaled,
+                   sgdnet::Family family,
+                   bool           fit_intercept,
+                   arma::uword    n_samples) {
+  double L;
+
+  switch(family) {
+    case sgdnet::GAUSSIAN: L = max_squared_sum + fit_intercept + alpha_scaled;
+                           break;
+  }
+
+  return 1.0 / (2.0*L + std::min(2.0*n_samples*alpha_scaled, L));
+}
+
+//' Return index of nonzero elements
+//'
+//' The function is overloaded to work with both dense and sparse matrices.
+//'
+//' @param x_it iterator
+//' @param i counter
+//'
+//' @return Index of iterator in current row or column.
+//'
+//' @keywords internal
+arma::uvec Nonzeros(const arma::sp_mat& x, arma::uword i) {
+
+  arma::sp_mat::const_col_iterator x_itr = x.begin_col(i);
+  arma::sp_mat::const_col_iterator x_end = x.end_col(i);
+
+  std::vector<unsigned int> out;
+
+  for(; x_itr != x_end; ++x_itr) {
+    out.push_back(x_itr.row());
+  }
+
+  return arma::conv_to< arma::uvec >::from(out);
+}
+
+// For dense matrices, just return a sequence of integers
+// TODO: consider using a conditional for sparse matrices to avoid
+//       entering this function.
+arma::uvec Nonzeros(const arma::mat& x, arma::uword i) {
+  return arma::regspace<arma::uvec>(0, x.n_rows - 1);
+}
+
+//' Preprocess data
+//'
+//' @param x feature matrix, sparse or dense
+//' @param y targets
+//' @param normalize whether to normalize x
+//' @param fit_intercept wheter to fit the intercept. If false, no
+//'   scaling or centering is done.
+//' @param x_offset a vector of offsets for each feature
+//' @param x_scale a vector of scaling factors (l2 norms) for each vector
+//' @param y_offset a vector of means for each column in y
+//'
+//' @return Nothing. x and y are scaled and centered.
+//' @keywords internal
+template <typename T>
+void Preprocess(T&            x,
+                arma::mat&    y,
+                const bool    normalize,
+                const bool    fit_intercept,
+                arma::vec&    x_offset,
+                arma::vec&    x_scale,
+                arma::rowvec& y_offset) {
+
+  // TODO: what is the reason for not scaling and centering when the intercept
+  //       is not fit?
+
+  if (fit_intercept) {
+
+    arma::uword n_features = x.n_rows;
+
+    // Center feature matrix
+    x_offset = arma::mean(x, 1);
+    for (arma::uword i = 0; i < n_features; ++i)
+      x.row(i) -= x_offset(i);
```

Comment:
This worries me a bit - if `x` is sparse and large, this will destroy the sparsity pattern (and blow out memory usage).

Reply:
Of course, you are right. So just stick to scaling when features are sparse?

Reply:
Scaling isn't the issue - centering is. 

For now, I think it's safe to throw an error, but long term we can probably support sparse X + centering with clever linear algebra. For example, the centered X, which I'll call X_c, can be written as X_c = X - m 1^T where m is the vector of column means and 1 is the vector of all ones. Then, e.g., we can replace X_c^Ty by `(X - m 1^T)^Ty = X^Ty - 1m^Ty` throughout the algorithm and work with X_c without ever materializing it. 

I don't think that's the right adjustment for our algorithm, but it should be possible to work something similar out. Once you write-up the technical / algorithmic docs, I'll take a look and try to work something out.

Reply:
Yes, sorry, that's what I meant. I suggest, however, that we just scale the features when the input is sparse and don't throw any errors.

Reply:
Errr.... I don't love the idea of the standardization being different for dense and sparse X. Even if we document it prominently, I would be really surprised if `sgdnet(X, y)` and `sgdnet(as.matrix(X), y)` gave different results...

Reply:
Well, if I am not mistaken, the only differences that emerge will result from differences in the relative size of the penalty. Since we are rescaling, we are otherwise going to get back the same coefficients. In any case, there are always going to be some discrepancies because of the differences in how the intercept is handled between the dense and sparse implementations.

```r
x <- Matrix::rsparsematrix(10000, 2, density = 0.5)
x[, 2] <- x[, 2]*100
x[x[, 2] != 0, 2] <- x[x[, 2] != 0, 2] + 10000
y <- x[, 1] + x[, 2]/100

set.seed(0)
coef1 <- coef(sgdnet(x, y, alpha = 0, lambda = 0))
coef2 <- coef(sgdnet(as.matrix(x), y, alpha = 0, lambda = 0))
cbind(coef1, coef2)

# 3 x 2 Matrix of class "dgeMatrix"
# s0          s0
# (Intercept) 0.185315314 0.069078897
# 1           1.005611802 0.993138237
# 2           0.009968166 0.009986274
```

It feels like there should be a way around this. glmnet returns equivalent results for sparse and dense matrices alike. I'll look into how they standardize their features in the sparse implementation.

Reply:
[Someone else already did the legwork on this](https://stackoverflow.com/questions/46835181/sklearn-perceptron-behaving-differently-for-sparse-and-dense) but they basically use a slower learning rate for the intercept in the sparse implementation to avoid intercept oscillation.

Reply:
That's interesting - ok, let's table this for now and possibly come back to it later. It's not worth holding up other development right now. 

Can you open an issue so we don't forget about it?

Reply:
Yes, good point. It would not be terribly hard to run some simulations and see how this plays out in our implementation.

[The issue has been filed](https://github.com/jolars/sgdnet/issues/2).